usr/src/uts/common/fs/zfs/dmu_zfetch.c

```diff
@@ -251,32 +256,48 @@ dmu_zfetch(zfetch_t *zf, uint64_t blkid, uint64_t nblks)
 	 * want to prefetch to block we just accessed.  In this case,
 	 * start just after the block we just accessed.
 	 */
-	int64_t pf_start = MAX(zs->zs_pf_blkid, blkid + nblks);
+	pf_start = MAX(zs->zs_pf_blkid, blkid + nblks);
 
 	/*
 	 * Double our amount of prefetched data, but don't let the
 	 * prefetch get further ahead than zfetch_max_distance.
 	 */
-	int pf_nblks =
-	    MIN((int64_t)zs->zs_pf_blkid - zs->zs_blkid + nblks,
-	    zs->zs_blkid + nblks +
-	    (zfetch_max_distance >> zf->zf_dnode->dn_datablkshift) - pf_start);
+	if (data) {
+		max_dist = zfetch_max_distance >> zf->zf_dnode->dn_datablkshift;
+		pf_nblks = MIN((int64_t)zs->zs_pf_blkid - blkid + nblks,
+		    blkid + nblks + max_dist - pf_start);
+	} else {
+		pf_nblks = 0;
+	}
 
 	zs->zs_pf_blkid = pf_start + pf_nblks;
-	zs->zs_atime = gethrtime();
-	zs->zs_blkid = blkid + nblks;
 
 	/*
-	 * dbuf_prefetch() issues the prefetch i/o
-	 * asynchronously, but it may need to wait for an
-	 * indirect block to be read from disk.  Therefore
-	 * we do not want to hold any locks while we call it.
+	 * Do the same for indirects, starting from where we stopped last,
+	 * or where we read data above (that implies also indirects).
 	 */
+	ipf_start = MAX(zs->zs_ipf_blkid, zs->zs_pf_blkid);
```

Comment:
Perhaps there aren't any invariants, in which case that's fine.  In general, what I mean is something like "If A is always (when entering this section of code) greater than B, then we should add an assertion to that effect".

In general, my issue with this code is that it's hard to read; perhaps adding assertions would help, or maybe there just need to be more intermediate variables with more descriptive names, or maybe there should just be some comments explaining what some of the longer calculations are doing.